Script portability improvements + fix for verify-randomly

fmoledina opened the pull request:
@alphapapa, up for consideration are the following changes:

* Included a `--config` command-line flag for specifying custom config location
* Included `restic` and `restic_opts` variables that can be specified in runner/repo/sets
* Updated the `verify-randomly` command to call restic and increased `sleep` timeout to 10 seconds for each attempt. This allowed for more successful mounting on my system.

My README.org includes a summary of the changes in the very top section, which providing a link to your original repo.

Looking forward to your input. Thanks!

fmoledina committed:
Merge pull request #1 from alphapapa/master

Update from upstream

fmoledina committed:
First commit

fmoledina committed:
First commit

alphapapa commented:
Hi,

Generally, I prefer that a PR not contain changes that are unrelated.  Especially, each commit should be atomic.

> Included a --config command-line flag for specifying custom config location

I'd probably be willing to add such an option, but the default should remain specified as it is.  So, e.g. the argument processing should change the `config_dir` variable when `--config` is specified.

> Included restic and restic_opts variables that can be specified in runner/repo/sets

I'd probably prefer not to add the `restic` variable.  The canonical way to handle paths in Unix scripts is to put the executable or a symlink to it in a directory in `PATH`.  So it's not necessary for the script to have this option.

The `restic_opts` variable might be a good idea.  Please look at some of the other, similar issues and PRs regarding passing additional options to Restic.

> Updated the verify-randomly command to call restic and increased sleep timeout to 10 seconds for each attempt. This allowed for more successful mounting on my system.

Increasing the sleep timeout is probably fine; or maybe it should be made an option.

Changing the command as you did is incorrect.  Note that `mount` is defined as a function in the script, one which makes additional correctness checks.

> Script portability improvements

I don't see any portability improvements in the PR.  Maybe the changes you did note are what you mean by that, but the term usually refers to changes which make programs function better on alternative platforms, like other OSes or other shell interpreters.

Thanks for sending in the PR.  I'd generally recommend opening issues to discuss improvements that are needed, and then if we agree on a solution, proposing a PR for each solution.

fmoledina commented:
Thanks for the prompt response and I welcome your commentary and guidance.

> Generally, I prefer that a PR not contain changes that are unrelated. Especially, each commit should be atomic.

Understood.

> I'd probably be willing to add such an option, but the default should remain specified as it is. So, e.g. the argument processing should change the `config_dir` variable when `--config` is specified.

This is the intended behavior. I changed the variable name from `config_dir` to `config` in order for it to play well with `getopt`, and declare the default value in line 549.

```bash
[[ $config ]] || config=~/.config/backup/restic
```

> I'd probably prefer not to add the `restic` variable. The canonical way to handle paths in Unix scripts is to put the executable or a symlink to it in a directory in `PATH`. So it's not necessary for the script to have this option.

Understood.

> The `restic_opts` variable might be a good idea. Please look at some of the other, similar issues and PRs regarding passing additional options to Restic.

I'll have a look at the other similar issues and PRs

> Increasing the sleep timeout is probably fine; or maybe it should be made an option.

> Changing the command as you did is incorrect. Note that `mount` is defined as a function in the script, one which makes additional correctness checks.

I think I confounded one change with another. I completely missed the fact that `mount` in this context is calling the defined function, which I also modified and tested ahead of working on `verify-randomly`. I conclude now that only increasing the `sleep` time was necessary for me. Not sure about other services, but I use Backblaze B2 and waiting for the mount would time out too fast.

What is the use of both the attempt countdown and the sleep time together? It is not as though the attempt loop is actually sending multiple mount attempts. Would it be better a single timeout instead with a configurable value?

> I don't see any portability improvements in the PR. Maybe the changes you did note are what you mean by that, but the term usually refers to changes which make programs function better on alternative platforms, like other OSes or other shell interpreters.

Agreed, I think I was looking for the word versatility, i.e. own config directory and custom restic options.

> Thanks for sending in the PR. I'd generally recommend opening issues to discuss improvements that are needed, and then if we agree on a solution, proposing a PR for each solution.

Fully understood on this and I'm ready to close this PR, and perhaps break out the `verify-randomly` question into a separate issue. Thank you!

alphapapa commented:
> What is the use of both the attempt countdown and the sleep time together?

Two reasons:

1.  It allows this debug info to log how long it actually took to mount:

```sh
            debug "Mounted repo not yet available ($attempts attempts): waiting..."
```

2.  If the sleep time were longer than it actually took to mount, the extra time would be wasted.  No need to wait 10 seconds if it only takes 2 to mount it.

> Fully understood on this and I'm ready to close this PR, and perhaps break out the verify-randomly question into a separate issue. Thank you!

Thanks.
